File: app/src/main/java/com/example/totoroto/mureok/PlantTip/TipActivity.java
package com.example.totoroto.mureok.PlantTip;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import com.example.totoroto.mureok.Data.TipData;
import com.example.totoroto.mureok.Data.TipDetailData;
import com.example.totoroto.mureok.R;
import com.example.totoroto.mureok.RecyclerItemClickListener;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserFactory;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;

public class TipActivity extends AppCompatActivity implements View.OnClickListener {
    private final String TAG = "TipPlant";
    public static final String INTENT_STR = "TipDetailData";

    private final String API_KEY = "20170726ADJPJHDALH0FXS5HK84T7G";
    private final String TAG_FRT = "frtlzrInfo";
    private final String TAG_TEMPERATURE = "grwhTpCode";
    private final String TAG_HYDRO = "hdCode";
    private final String TAG_PRPG = "prpgtEraInfo";
    private final String TAG_SOIL = "soilInfo";
    private final String TAG_WATER_SPRING = "watercycleSprngCode";
    private final String TAG_WATER_SUMMER = "watercycleSummerCode";
    private final String TAG_WATER_AUTUMN = "watercycleAutumnCode";
    private final String TAG_WATER_WINTER = "watercycleWinterCode";

    private final int numOfRows = 216;
    private ArrayList<TipData> tempLists;
    private ArrayList<TipData> mTipDatas;
    private ArrayList<TipData> filteredLists;
    private RecyclerView recyclerTip;
    private TipAdapter tipAdapter;
    private LinearLayoutManager layoutManager;
    private Button btnBack;
    private EditText etTipSearch;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_tip);

        init();
        aboutRecycler();
        aboutTipDetail();
        new Thread(new Runnable() {
            @Override
            public void run() {
                getPlantListData();

                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        aboutMapListData();
                    }
                });
            }
        }).start();

        aboutEditTextSearch();
    }

    private void aboutEditTextSearch() {
        etTipSearch.addTextChangedListener(new TextWatcher() {

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                filteredLists.clear();
                s = s.toString().toLowerCase();

                for(int i=0; i<mTipDatas.size(); i++){
                    String str = mTipDatas.get(i).getpRealName(); //식물 이름
                    if(str.contains(s)){
                        filteredLists.add(mTipDatas.get(i));
                    }
                }
                tipAdapter.setTipDatas(filteredLists);
                tipAdapter.notifyDataSetChanged();
            }
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }
            @Override
            public void afterTextChanged(Editable s) {
            }
        });

    }

    private void aboutTipDetail() {

        recyclerTip.addOnItemTouchListener(new RecyclerItemClickListener(getApplicationContext(),
                recyclerTip, new RecyclerItemClickListener.OnItemClickListener() {
            @Override
            public void onItemClick(View view, final int position) {
                final int plantCode;
                if(filteredLists.size() != 0) {
                    plantCode = filteredLists.get(position).getpCode();
                    Log.d(TAG, "pCode" + plantCode + "|position" + position);
                }else{
                    plantCode = mTipDatas.get(position).getpCode();
                    Log.d(TAG, "pCode" + plantCode + "|position" + position);
                }
                new Thread(new Runnable() {
                    @Override
                    public void run() {
                        getPlantDetailData(plantCode, position);
                    }
                }).start();

            }
        }));
    }

    private void getPlantDetailData(int pCode, int position) {
        //1.
        String queryUrl = "http://api.nongsaro.go.kr/service/garden/gardenDtl"
                + "?apiKey=" + API_KEY + "&cntntsNo=" + pCode;

        try {
            URL url = new URL(queryUrl); //문자열로 된 요청 url을 URL 객체로 생성.
            InputStream is = url.openStream();  //url위치로 입력스트림 연결

            XmlPullParserFactory factory = XmlPullParserFactory.newInstance();
            XmlPullParser xpp = factory.newPullParser();

            xpp.setInput(new InputStreamReader(is, "UTF-8"));  //inputstream 으로부터 xml 입력받기

            String tag;

            xpp.next();
            int eventType = xpp.getEventType();
            TipDetailData tipDetailData = new TipDetailData();

            while (eventType != XmlPullParser.END_DOCUMENT) {
                switch (eventType) {
                    case XmlPullParser.START_TAG:
                        tag = xpp.getName();    //태그 이름 얻어오기

                        if (tag.equals("item")) ;
                        else if (tag.equals(TAG_FRT)) {
                            xpp.next();
                            tipDetailData.frt = xpp.getText();
                            Log.d(TAG, tipDetailData.frt);
                        }else if (tag.equals(TAG_TEMPERATURE)) {
                            xpp.next();
                            tipDetailData.temperature = xpp.getText();
                        }else if (tag.equals(TAG_HYDRO)) {
                            xpp.next();
                            tipDetailData.hydro = xpp.getText();
                        }else if (tag.equals(TAG_PRPG)) {
                            xpp.next();
                            tipDetailData.prpg = xpp.getText();
                        }else if (tag.equals(TAG_SOIL)) {
                            xpp.next();
                            tipDetailData.soil = xpp.getText();
                        }else if (tag.equals(TAG_WATER_SPRING)) {
                            xpp.next();
                            tipDetailData.waterSpring = xpp.getText();
                        }else if (tag.equals(TAG_WATER_SUMMER)) {
                            xpp.next();
                            tipDetailData.waterSummer = xpp.getText();
                        }else if (tag.equals(TAG_WATER_AUTUMN)) {
                            xpp.next();
                            tipDetailData.waterAutumn = xpp.getText();
                        }else if (tag.equals(TAG_WATER_WINTER)) {
                            xpp.next();
                            tipDetailData.waterWinter = xpp.getText();
                        }
                        break;
                }
                eventType = xpp.next();

                if(tipDetailData.waterWinter != null) {
                    eventType = XmlPullParser.END_DOCUMENT; //while 종료


                    if(filteredLists.size() == 0) {
                        //detail 액티비티로 객체 보내기
                        Intent intent = new Intent(getApplicationContext(), TipDetailActivity.class);
                        intent.putExtra(INTENT_STR, tipDetailData);
                        intent.putExtra("plantImage", mTipDatas.get(position).getpImage());
                        intent.putExtra("plantName", mTipDatas.get(position).getpRealName());
                        startActivity(intent);
                    }else{
                        Intent intent = new Intent(getApplicationContext(), TipDetailActivity.class);
                        intent.putExtra(INTENT_STR, tipDetailData);
                        intent.putExtra("plantImage", filteredLists.get(position).getpImage());
                        intent.putExtra("plantName", filteredLists.get(position).getpRealName());
                        startActivity(intent);
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void init() {
        recyclerTip = (RecyclerView)findViewById(R.id.recyclerTip);
        btnBack = (Button)findViewById(R.id.btnBack_tip);
        etTipSearch = (EditText)findViewById(R.id.etTipSearch);

        btnBack.setOnClickListener(this);
        mTipDatas = new ArrayList<>();
        filteredLists = new ArrayList<>();
    }

    private void aboutRecycler() {
        layoutManager = new LinearLayoutManager(getApplicationContext());
        recyclerTip.setHasFixedSize(true);
        recyclerTip.setLayoutManager(layoutManager);
        DividerItemDecoration mDividerItemDeco = new DividerItemDecoration(
                recyclerTip.getContext(), layoutManager.getOrientation());
        recyclerTip.addItemDecoration(mDividerItemDeco);

        tipAdapter = new TipAdapter();
        tipAdapter.setTipDatas(mTipDatas);
        recyclerTip.setAdapter(tipAdapter);
    }



    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btnBack_tip:
                finish();
                break;
        }
    }

    public void getPlantListData() {
        tempLists = new ArrayList<>();
        int pNo = 0;
        String pName = "";
        String pImage = "";

        String queryUrl = "http://api.nongsaro.go.kr/service/garden/gardenList"
                + "?apiKey=" + API_KEY + "&numOfRows=" + numOfRows;

        try {
            URL url = new URL(queryUrl); //문자열로 된 요청 url을 URL 객체로 생성.
            InputStream is = url.openStream();  //url위치로 입력스트림 연결

            XmlPullParserFactory factory = XmlPullParserFactory.newInstance();
            XmlPullParser xpp = factory.newPullParser();

            xpp.setInput(new InputStreamReader(is, "UTF-8"));  //inputstream 으로부터 xml 입력받기

            String tag;
            xpp.next();
            int eventType = xpp.getEventType();

            while (eventType != XmlPullParser.END_DOCUMENT) {
                switch (eventType) {
                    case XmlPullParser.START_TAG:
                        tag = xpp.getName();    //태그 이름 얻어오기

                        if (tag.equals("item")) ;
                        else if (tag.equals("cntntsNo")) { //식물 넘버
                            xpp.next();
                            pNo = Integer.valueOf(xpp.getText());
                        } else if (tag.equals("cntntsSj")) { //식물명
                            xpp.next();
                            pName = xpp.getText();
                        }else if(tag.equals("rtnStreFileNm")){ //이미지 파일 경로
                            xpp.next();
                            pImage = xpp.getText();
                            TipData data = new TipData(pNo, pName, pImage);
                            tempLists.add(data);
                        }

                        break;
                }
                eventType = xpp.next();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
//
    }

    private void aboutMapListData() {
        for(int i=0; i<tempLists.size(); i++){
            TipData tData = new TipData(tempLists.get(i).getpCode(), tempLists.get(i).getpRealName(), tempLists.get(i).getpImage());
            mTipDatas.add(tData);
        }

        tipAdapter.notifyDataSetChanged();
    }
}

File: app/src/main/java/com/example/totoroto/mureok/PlantTip/TipActivity.kt
package com.example.totoroto.mureok.PlantTip

import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import com.example.totoroto.mureok.Data.DetailParams
import com.example.totoroto.mureok.Data.TipData
import com.example.totoroto.mureok.Data.TipDetailData
import com.example.totoroto.mureok.R
import com.example.totoroto.mureok.RecyclerItemClickListener
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory
import java.io.InputStreamReader
import java.net.URL

class TipActivity : AppCompatActivity(), View.OnClickListener {

    private var tempLists: ArrayList<TipData>? = null
    private var filteredLists: ArrayList<TipData> = ArrayList()
    private var tipDatas: ArrayList<TipData> = ArrayList()

    private lateinit var tipAdapter: TipAdapter
    private lateinit var layoutManager: LinearLayoutManager
    private lateinit var btnBack: Button
    private lateinit var etTipSearch: EditText
    private lateinit var recyclerTip: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_tip)

        init()
        aboutRecycler()
        aboutTipDetail()

        val loadTask = LoadTask()
        loadTask.execute()

        aboutEditTextSearch()
    }

    private inner class LoadTask : AsyncTask<Void, Void, Void?>() {
        override fun doInBackground(vararg voids: Void): Void? {
            getPlantListData()
            return null
        }

        override fun onPostExecute(aVoid: Void?) {
            super.onPostExecute(aVoid)
            aboutMapListData()
        }
    }


    private fun aboutEditTextSearch() {
        etTipSearch.addTextChangedListener(object : TextWatcher {

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                var s = s
                filteredLists.clear()
                s = s.toString().toLowerCase()

                for (i in 0..tipDatas.size) {
                    val str = tipDatas.get(i).realName //식물 이름
                    if (str?.contains(s) != null) {
                        filteredLists.add(tipDatas.get(i))
                    }
                }
                tipAdapter.setTipDatas(filteredLists)
                tipAdapter.notifyDataSetChanged()
            }

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun afterTextChanged(s: Editable) {}
        })

    }

    private fun aboutTipDetail() {

        recyclerTip.addOnItemTouchListener(RecyclerItemClickListener(applicationContext,
                recyclerTip, RecyclerItemClickListener.OnItemClickListener { view, position ->
            val plantCode: Int?

            if (filteredLists.size != 0) {
                plantCode = filteredLists.get(position).code
                Log.d(TAG, "pCode$plantCode|position$position")
            } else {
                plantCode = tipDatas.get(position).code
                Log.d(TAG, "pCode$plantCode|position$position")
            }

            val detailParams = DetailParams(plantCode, position);
            val detailTask = DetailTask()
            detailTask.execute(detailParams)
        }))
    }

    private inner class DetailTask : AsyncTask<DetailParams, Void, Void?>{
        constructor() : super()

        override fun doInBackground(vararg params: DetailParams): Void? {
            getPlantDetailData(params[0].plantCode, params[0].position)
            return null
        }
    }

    private fun getPlantDetailData(pCode: Int?, position: Int) {
        val queryUrl = ("http://api.nongsaro.go.kr/service/garden/gardenDtl"
                + "?apiKey=" + API_KEY + "&cntntsNo=" + pCode)

        try {
            val url = URL(queryUrl) //문자열로 된 요청 url을 URL 객체로 생성.
            val `is` = url.openStream()  //url위치로 입력스트림 연결

            val factory = XmlPullParserFactory.newInstance()
            val xpp = factory.newPullParser()

            xpp.setInput(InputStreamReader(`is`, "UTF-8"))  //inputstream 으로부터 xml 입력받기

            var tag: String

            xpp.next()
            var eventType = xpp.eventType
            val tipDetailData = TipDetailData()

            while (eventType != XmlPullParser.END_DOCUMENT) {
                when (eventType) {
                    XmlPullParser.START_TAG -> {
                        tag = xpp.name    //태그 이름 얻어오기

                        if (tag == "item")
                        else if (tag == TAG_FRT) {
                            xpp.next()
                            tipDetailData.frt = xpp.text
                        } else if (tag == TAG_TEMPERATURE) {
                            xpp.next()
                            tipDetailData.temperature = xpp.text
                        } else if (tag == TAG_HYDRO) {
                            xpp.next()
                            tipDetailData.hydro = xpp.text
                        } else if (tag == TAG_PRPG) {
                            xpp.next()
                            tipDetailData.prpg = xpp.text
                        } else if (tag == TAG_SOIL) {
                            xpp.next()
                            tipDetailData.soil = xpp.text
                        } else if (tag == TAG_WATER_SPRING) {
                            xpp.next()
                            tipDetailData.waterSpring = xpp.text
                        } else if (tag == TAG_WATER_SUMMER) {
                            xpp.next()
                            tipDetailData.waterSummer = xpp.text
                        } else if (tag == TAG_WATER_AUTUMN) {
                            xpp.next()
                            tipDetailData.waterAutumn = xpp.text
                        } else if (tag == TAG_WATER_WINTER) {
                            xpp.next()
                            tipDetailData.waterWinter = xpp.text
                        }
                    }
                }
                eventType = xpp.next()

                if (tipDetailData.waterWinter != null) {
                    eventType = XmlPullParser.END_DOCUMENT //while 종료


                    if (filteredLists.size == 0) {
                        //detail 액티비티로 객체 보내기
                        val intent = Intent(applicationContext, TipDetailActivity::class.java)
                        intent.putExtra(INTENT_STR, tipDetailData)
                        intent.putExtra("plantImage", tipDatas.get(position).image)
                        intent.putExtra("plantName", tipDatas.get(position).realName)
                        startActivity(intent)
                    } else {
                        val intent = Intent(applicationContext, TipDetailActivity::class.java)
                        intent.putExtra(INTENT_STR, tipDetailData)
                        intent.putExtra("plantImage", filteredLists.get(position).image)
                        intent.putExtra("plantName", filteredLists.get(position).realName)
                        startActivity(intent)
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun init() {
        recyclerTip = findViewById<View>(R.id.recyclerTip) as RecyclerView
        btnBack = findViewById<View>(R.id.btnBack_tip) as Button
        etTipSearch = findViewById<View>(R.id.etTipSearch) as EditText

        btnBack.setOnClickListener(this)
    }

    private fun aboutRecycler() {
        layoutManager = LinearLayoutManager(applicationContext)
        recyclerTip.setHasFixedSize(true)
        recyclerTip.layoutManager = layoutManager
        val mDividerItemDeco = DividerItemDecoration(recyclerTip.context, layoutManager.orientation)
        recyclerTip.addItemDecoration(mDividerItemDeco)

        tipAdapter = TipAdapter()
        tipAdapter.setTipDatas(tipDatas)
        recyclerTip.adapter = tipAdapter
    }


    override fun onClick(v: View) {
        when (v.id) {
            R.id.btnBack_tip -> finish()
        }
    }

    fun getPlantListData() {
        tempLists = ArrayList()
        var pNo = 0
        var pName = ""
        var pImage = ""

        val queryUrl = ("http://api.nongsaro.go.kr/service/garden/gardenList"
                + "?apiKey=" + API_KEY + "&numOfRows=" + NUM_OF_ROWS)

        try {
            val url = URL(queryUrl) //문자열로 된 요청 url을 URL 객체로 생성.
            val `is` = url.openStream()  //url위치로 입력스트림 연결

            val factory = XmlPullParserFactory.newInstance()
            val xpp = factory.newPullParser()

            xpp.setInput(InputStreamReader(`is`, "UTF-8"))  //inputstream 으로부터 xml 입력받기

            var tag: String
            xpp.next()
            var eventType = xpp.eventType

            while (eventType != XmlPullParser.END_DOCUMENT) {
                when (eventType) {
                    XmlPullParser.START_TAG -> {
                        tag = xpp.name    //태그 이름 얻어오기

                        if (tag == "item")
                        else if (tag == "cntntsNo") { //식물 넘버
                            xpp.next()
                            pNo = Integer.valueOf(xpp.text)
                        } else if (tag == "cntntsSj") { //식물명
                            xpp.next()
                            pName = xpp.text
                        } else if (tag == "rtnStreFileNm") { //이미지 파일 경로
                            xpp.next()
                            pImage = xpp.text
                            val data = TipData(pNo, pName, pImage)
                            tempLists?.add(data)
                        }
                    }
                }
                eventType = xpp.next()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun aboutMapListData() {
        val listSize = tempLists?.size ?: return

        for (i in 0 until listSize) {
            val tData = TipData(tempLists?.get(i)?.code, tempLists?.get(i)?.realName, tempLists?.get(i)?.image)
            tipDatas.add(tData)
        }

        tipAdapter.notifyDataSetChanged()
    }

    companion object {
        val INTENT_STR = "TipDetailData"
        private val TAG = "TipPlant"

        private val API_KEY = "20170726ADJPJHDALH0FXS5HK84T7G"
        private val TAG_FRT = "frtlzrInfo"
        private val TAG_TEMPERATURE = "grwhTpCode"
        private val TAG_HYDRO = "hdCode"
        private val TAG_PRPG = "prpgtEraInfo"
        private val TAG_SOIL = "soilInfo"
        private val TAG_WATER_SPRING = "watercycleSprngCode"
        private val TAG_WATER_SUMMER = "watercycleSummerCode"
        private val TAG_WATER_AUTUMN = "watercycleAutumnCode"
        private val TAG_WATER_WINTER = "watercycleWinterCode"

        private val NUM_OF_ROWS = 216
    }
}


